Carbon.Cryptography/Functions/Convert-CCertificateProvider.ps1

```diff
@@ -0,0 +1,161 @@
+    * `NewCertificateBase64Encoded`: the raw bytes of the new certificate file, base-64 encoded.
+
+    The `certutil -csplist` shows a list of available cryptographic providers.
+
+    .EXAMPLE
+    Convert-CCertificateProvider -FilePath .\mycert.pfx -ProviderName 'Microsoft Enhanced RSA and AES Cryptographic Provider'
+
+    Demonstrates how to convert the provider of a certificate's private key and the certificate file is ***not***
+    password protected.
+
+    .EXAMPLE
+    Convert-CCertificateProvider -FilePath .\mycert.pfx -ProviderName 'Microsoft Enhanced RSA and AES Cryptographic Provider'
+
+    Demonstrates how to convert the provider of a certificate's private key and the certificate file ***is*** password
+    protected.
+    #>
+    [CmdletBinding()]
+    param(
+        # The path to the certificate file to convert. Must have a private key.
+        [Parameter(Mandatory)]
+        [String] $FilePath,
+
+        # The new provider name for the certifcate's private key. The `certutil -csplist` command shows the list of
+        # available cryptographic providers.
+        [Parameter(Mandatory)]
+        [String] $ProviderName,
+
+        # The password for the certificate file, if any. When replacing the existing certificate file, it will be
+        # protected with the same password (or not protected if no password is passed).
+        [securestring] $Password
+    )
+
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+
+    if (-not (Test-Path -Path $FilePath -PathType Leaf))
+    {
+        "Unable to convert provider for certificate ""$($FilePath)"" because it does not exist." |
+            Write-Error -ErrorAction $ErrorActionPreference
+        return
+    }
+
+    $FilePath = $FilePath | Resolve-Path | Select-Object -ExpandProperty 'ProviderPath'
+
+    $pwdArg = @{}
+    if ($Password)
+    {
+        $pwdArg['Password'] = $Password
+    }
+    $cert = Get-CCertificate -Path $FilePath @pwdArg
+
+    if (-not $cert.PrivateKey)
+    {
+        "Unable to convert provider for certificate ""$($FilePath)"" because it does not have a private key." |
+            Write-Error -ErrorAction $ErrorActionPreference
+        return
+    }
+
+    $pk = $cert.PrivateKey
+    $pkProviderName = ''
+    if ($pk | Get-Member 'Key')
+    {
+        $pkProviderName = $pk.Key.Provider.Provider
+    }
+    elseif ($pk | Get-Member 'CspKeyContainerInfo')
+    {
+        $pkProviderName = $pk.CspKeyContainerInfo.ProviderName
+    }
+    else
+    {
+        "Unable to convert provider for certificate ""$($FilePath)"" because it does not have a supported private key " +
+            'implementation.' | Write-Error -ErrorAction $ErrorActionPreference
+        return
+    }
+
+    if ($pkProviderName -eq $ProviderName)
+    {
+        Write-Verbose "Certificate ""$($FilePath)"" private key already using provider ""$($pkProviderName)""."
+        return

```

Comment:
The function isn't returning the expected object here and not notifying the user as to why (the verbose message doesn't count because that stream is silent by default). In that case we should either write a warning and return nothing or return the same object as when the provider is changed.